nscplugin/src/main/scala/scala/scalanative/nscplugin/NirCodeGen.scala

```diff
@@ -1540,12 +1540,27 @@ abstract class NirCodeGen
           val offset = genExpr(offsetp, ptr)
           offset withOp Op.Elem(ty, ptr.value, Seq(offset.value))
 
-        case (PTR_SUB, Seq(offsetp, tagp)) =>
-          val st     = unwrapTag(tagp)
-          val ty     = genType(st, box = false)
-          val offset = genExpr(offsetp, ptr)
-          val neg    = negateInt(offset.value, offset)
-          neg withOp Op.Elem(ty, ptr.value, Seq(neg.value))
+        case (PTR_SUB, Seq(argp, tagp)) =>
+          val st  = unwrapTag(tagp)
+          val ty  = genType(st, box = false)
+          val sym = argp.tpe.typeSymbol
+          sym match {
+            case IntClass =>
+              val offset = genExpr(argp, ptr)
+              val neg    = negateInt(offset.value, offset)
+              neg withOp Op.Elem(ty, ptr.value, Seq(neg.value))
+
+            case PtrClass =>
+              // Pointers in Scala Native are untyped and modeled as `i8*`.
+              // Pointer substraction therefore explicitly divide the byte
+              // offset by the size of pointer type.
+              val ptrInt     = ptr withOp Op.Conv(nir.Conv.Ptrtoint, nir.Type.ULong, ptr.value)
+              val ptrArg     = genExpr(argp, ptrInt)
+              val ptrArgInt  = ptrArg withOp Op.Conv(nir.Conv.Ptrtoint, nir.Type.ULong, ptrArg.value)
```

Comment:
Done. I don't know if it would make sense to also resolve the type of `CPtrDiff` using symbols or whether it is OK to explicitly say `Long` when calculating the result type.